src/adagrams.js

```diff
@@ -1,15 +1,189 @@
+    let letter = lettersInHand[idx];
+    if (!(letter in letterDict)) {
+      letterDict[letter] = 1;
+    } else {
+      letterDict[letter]++;
+    }
+  }
+  const word = input.toUpperCase();
+  let keepChecking = true;
+  for (let idx = 0; idx < word.length; idx++) {
+    let letter = word[idx];
+    if (!(letter in letterDict)) {
+      keepChecking = false;
+      break;
+    } else {
+      letterDict[letter] -= 1;
+      if (letterDict[letter] < 0) {
+        keepChecking = false;
+        break;
+      }
+    }
+  }
+  return keepChecking;
 };
 
 export const scoreWord = (word) => {
   // Implement this method for wave 3
+  if (word.length === 0) {
+    return 0;
+  }
+  const scoreChart = {
+    A: 1,
+    B: 3,
+    C: 3,
+    D: 2,
+    E: 1,
+    F: 4,
+    G: 2,
+    H: 4,
+    I: 1,
+    J: 8,
+    K: 5,
+    L: 1,
+    M: 3,
+    N: 1,
+    O: 1,
+    P: 3,
+    Q: 10,
+    R: 1,
+    S: 1,
+    T: 1,
+    U: 1,
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10,
+  };
+  let scorePoint = 0;
+  word = word.toUpperCase();
+  for (let idx = 0; idx < word.length; idx++) {
+    let letter = word[idx];
+    scorePoint += scoreChart[letter];
+  }
+  if (word.length >= 7) {
+    scorePoint += 8;
+  }
+  return scorePoint;
 };
 
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 1
+  // Implement this method for wave 4
+  //Create list of dictionaries storing each word and its score
+  const scoreslist = [];
+  for (let idx = 0; idx < words.length; idx++) {
+    let word = words[idx];
+    //calling on the previous scoreWord function
+    let score = scoreWord(word);
+    scoreslist.push({
+      word: word,
+      score: score,
+    });
+  }
+
+  //Create list of dictionaries storing word and score for top-scoring word(s)
+  const topWords = [];
+  let max = scoreslist[0]["score"];
+  for (let idx = 1; idx < scoreslist.length; idx++) {
+    let word = scoreslist[idx];
+    if (scoreslist[idx]["score"] > max) {
+      max = scoreslist[idx]["score"];
+    }
+  }
+  for (let idx = 0; idx < scoreslist.length; idx++) {
+    if (scoreslist[idx]["score"] === max) {
+      topWords.push(scoreslist[idx]);
+    }
+  }
+  //# If there is only one top word, return info for that one
+  if (topWords.length === 1) {
+    return topWords[0];
+  } else if (topWords.length < 1) {
+    return null;
+  } else if (topWords.length > 1) {
+    // See if there is a word of length 10
+    for (let idx = 0; idx < topWords.length; idx++) {
+      if (topWords[idx]["word"].length === 10) {
+        return topWords[idx];
+      }
+    }
+    let min = topWords[0]["word"].length;
+    for (let idx = 1; idx < topWords.length; idx++) {
+      if (topWords[idx]["word"].length < min) {
+        min = topWords[idx]["word"].length;
+      }
+    }
+    for (let idx = 0; idx < topWords.length; idx++) {
+      if (topWords[idx]["word"].length === min) {
+        return topWords[idx];
+      }
+    }
+  }
```

Comment:
I think we can combine a lot of this together. Instead of tracking every word and its score, we only want to track the current winning word and its score. We could create one loop that scores a word, checks it against the current highest score.

If the new word is scored higher, it becomes the current winning word. If the word has the same score as the current winning word, we check its length.

Once the for loop is done, we return the winning word